src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S

```diff
@@ -532,3 +532,146 @@ NESTED_ENTRY RhpCallFilterFunclet, _TEXT, NoHandler
         ret
 
 NESTED_END RhpCallFilterFunclet, _TEXT
+
+
+#ifdef FEATURE_OBJCMARSHAL
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////
+//
+// void* FASTCALL RhpCallPropagateExceptionCallback(void* pCallbackContext, void* pCallback, REGDISPLAY* pRegDisplay,
+//                                    ExInfo* pExInfo)
+//
+// INPUT:  RDI:  callback context
+//         RSI:  callback
+//         RDX:  REGDISPLAY*
+//         RCX:  ExInfo*
+//
+// OUTPUT:
+//
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////
+NESTED_ENTRY RhpCallPropagateExceptionCallback, _TEXT, NoHandler
```

Comment:
Where does the switch to preemptive mode happens as we transition out of the managed code? Do we need to switch to preemptive mode somewhere in this method?

Reply:
I believe the thread state flag `TSF_DoNotTriggerGc` corresponds to cooperative mode, correct? It is cleared before restoring the registers:

```asm
// Clear the DoNotTriggerGc state before calling out to the propagation callback.
lock and     dword ptr [rax + OFFSETOF__Thread__m_ThreadStateFlags], ~TSF_DoNotTriggerGc
```

https://github.com/dotnet/runtime/pull/80334/files/d2b5cd8a1807577f4895e44f77eb4fa1a3ea17f4#diff-6d87d47bde66f3bf88ea6a7e87f5ceba784e8f02024ef1b7beda7a6ad7ba50a6R575-R576

Reply:
`TSF_DoNotTriggerGc` does not indicate preemptive/cooperative mode.

The thread is in preemptive mode when `Thread::m_pTransitionFrame` is non-null.

You can check for problems by calling `[UnmanagedCallersOnly]` callback from the unmanaged propagation function. If the mode is not switched properly, you will hit `Invalid Program: attempted to call a UnmanagedCallersOnly method from managed code` failfast.

Reply:
`TSF_DoNotTriggerGc` makes a thread uninterruptible/unhijackable when it runs managed code. I.E. managed code called form GC should not initiate or try to stop for GC as that would deadlock, so GC callouts set this flag.

The use of  `TSF_DoNotTriggerGc` in exception handling is less clear. I found few comments on it and they are a bit confusing  - 
https://github.com/dotnet/runtime/blob/b2dc37ba181a7fa4427e717eab819ba3543d0ae4/src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.asm#L543

The idea sems to be to disallow hijacking once the stack is inconsistent state due to exception unwind. Indeed we could have a hijack in the portion of the stack that just went dead.  
It looks like we permit hijacking in funclets, but not "between" funclets.

The flag certainly does not switch the thread to preemptive mode.

Reply:
More relevant comment - 
https://github.com/dotnet/runtime/blob/8918adf2e5255920aed3c61f9b92c2a51eec0fe3/src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs#L672

We do not allow GC in managed EH dispatch code once we are in the second pass. 
GC in funclets is ok though, thus it is toggled when calling the funclets.

Reply:
If I understand this correctly - we are done with exception dispatching and are basically passing the control to a native handler while the managed frames that were reverse-pinvoked and faulted are effectively gone. 
If GC happens, it will need to scan the current stack starting from some higher managed frame, if there was one, that called the native code, that called the managed code which faulted.

Basically we need to fetch the innermost pinvoke frame (`m_savedPInvokeTransitionFrame`) and put it in the `m_pTransitionFrame` - the same way as return from reverse pinvoke does. It could be `TOP_OF_STACK_MARKER` if there are no more managed frames.

`TSF_DoNotTriggerGc` should be cleared too, since we are done dispatching.

Reply:
Ah, I think I see what you are saying. Since this unwinds pass the reverse P/Invoke frame, it never invokes `RhpReversePInvokeReturn`, so the thread won't call this to return to preemptive mode:

https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/Runtime/thread.cpp#L1234-L1238

The `StackFrameIterator` does find the `PInvokeTransitionFrame` when unwinding, so perhaps I could restore that that to re-enable preemptive mode?

Reply:
Yes, that should work.

Reply:
Done. I re-ran the tests on arm64 and x64 (well, Rosetta 2 technically) to confirm that the augmented test now passes.